python/cog/server/helpers.py

```diff
@@ -1,28 +1,38 @@
+from __future__ import annotations
+
+import contextlib
 import io
 import os
 import selectors
+import sys
 import threading
 import uuid
-from typing import Callable, Optional, Sequence, TextIO
+from types import TracebackType
+from typing import Callable, Sequence, TextIO
+
+from typing_extensions import Self
 
 
-class WrappedStream:
+class _StreamWrapper:
     def __init__(self, name: str, stream: TextIO) -> None:
         self.name = name
         self._stream = stream
-        self._original_fp: Optional[TextIO] = None
-        self._wrapped_fp: Optional[TextIO] = None
+        self._original_fp: TextIO | None = None
+        self._wrapped_fp: TextIO | None = None
 
     def wrap(self) -> None:
+        if self._wrapped_fp or self._original_fp:
+            raise RuntimeError("stream is already wrapped")
+
         r, w = os.pipe()
 
-        # Save a copy of the original stream file descriptor.
+        # Save a copy of the stream file descriptor.
         original_fd = self._stream.fileno()
         original_fd_copy = os.dup(original_fd)
 
-        # Overwrite the original file descriptor with the write end of the
-        # pipe.
-        os.dup2(w, original_fd)
+        # Overwrite the stream file descriptor with the write end of the pipe.
+        os.dup2(w, self._stream.fileno())
+        os.close(w)
```

Comment:
This `close` fixes a file descriptor leak in the existing code, and has an associated test (`test_stream_redirector_does_not_leak_file_descriptors`) that fails if this is not here.